homeassistant/components/integration/sensor.py

```diff
@@ -244,15 +312,42 @@ def __init__(
         self._last_valid_state: Decimal | None = None
         self._attr_device_info = device_info
 
-    def _unit(self, source_unit: str) -> str:
-        """Derive unit from the source sensor, SI prefix and time unit."""
+    def _integrate_unit_over_time(self, source_unit: str) -> str:
         unit_time = self._unit_time_str
         if source_unit.endswith(f"/{unit_time}"):
             integral_unit = source_unit[0 : (-(1 + len(unit_time)))]
         else:
             integral_unit = f"{source_unit}{unit_time}"
 
-        return self._unit_template.format(integral_unit)
+        return integral_unit
+
+    def _derive_and_set_attributes_from_state(self, source_state: State) -> None:
+        source_unit = source_state.attributes.get(ATTR_UNIT_OF_MEASUREMENT)
+        if (
+            source_unit is not None
+        ):  # If the source has no defined unit we cannot derive a unit for the integral
+            self._unit_of_measurement = self._unit_template.format(
```

Comment:
I would keep this in the function that handles the unit... (as it was)

Reply:
when not handle all things related with the unit in a single function ?

Reply:
Ok. Here it is https://github.com/home-assistant/core/pull/113932/commits/b2e2299d42c923ce19b323af2ba3ec661c28e0b3 . I also moved the prefix logic from the constructor to that function.

Reply:
still think a single function to handle all thing related to unit called *_unit* was a better solution 😅  

Leave it to @emontnemery to decide